cloud/services/virtualmachines/virtualmachines.go

```diff
@@ -211,6 +200,91 @@ func (s *Service) Delete(ctx context.Context, spec interface{}) error {
 	return err
 }
 
+// generateStorageProfile generates a pointer to a compute.StorageProfile which can utilized for VM creation.
+func generateStorageProfile(vmSpec Spec) (*compute.StorageProfile, error) {
+	// TODO: Validate parameters before building storage profile
+	storageProfile := &compute.StorageProfile{
+		OsDisk: &compute.OSDisk{
+			Name:         to.StringPtr(azure.GenerateOSDiskName(vmSpec.Name)),
+			OsType:       compute.OperatingSystemTypes(vmSpec.OSDisk.OSType),
+			CreateOption: compute.DiskCreateOptionTypesFromImage,
+			DiskSizeGB:   to.Int32Ptr(vmSpec.OSDisk.DiskSizeGB),
+			ManagedDisk: &compute.ManagedDiskParameters{
+				StorageAccountType: compute.StorageAccountTypes(vmSpec.OSDisk.ManagedDisk.StorageAccountType),
+			},
+		},
+	}
+
+	imageRef, err := generateImageReference(vmSpec.Image)
+	if err != nil {
+		return nil, err
+	}
+
+	storageProfile.ImageReference = imageRef
+
+	return storageProfile, nil
+}
+
+// generateImageReference generates a pointer to a compute.ImageReference which can utilized for VM creation.
+func generateImageReference(image infrav1.Image) (*compute.ImageReference, error) {
+	imageRef := &compute.ImageReference{}
+
+	if image.ID != nil {
+		imageRef.ID = to.StringPtr(*image.ID)
+
+		// return early since we should only need the image ID
+		return imageRef, nil
+	} else if image.SubscriptionID != nil && image.ResourceGroup != nil && image.Gallery != nil && image.Name != nil && image.Version != nil {
+		imageID, err := generateImageID(image)
+		if err != nil {
+			return nil, err
+		}
+
+		imageRef.ID = to.StringPtr(imageID)
+
+		// return early since we're referencing an image that may not be published
+		return imageRef, nil
+	}
+
+	if image.Publisher != nil {
+		imageRef.Publisher = image.Publisher
+	}
+	if image.Offer != nil {
+		imageRef.Offer = image.Offer
+	}
+	if image.SKU != nil {
+		imageRef.Sku = image.SKU
+	}
+	if image.Version != nil {
+		imageRef.Version = image.Version
+
+		return imageRef, nil
+	}
+
+	return nil, errors.Errorf("Image reference cannot be generated, as fields are missing: %+v", *imageRef)
+}
+
+// generateImageID generates the resource ID for an image stored in an Azure Shared Image Gallery.
+func generateImageID(image infrav1.Image) (string, error) {
```

Comment:
Don't need to address in this PR, but we should probably enable webhook validation so clients can get errors right away instead of at reconcile time.

Reply:
@juan-lee -- Would you mind opening an issue for webhook validation and linking back to this comment?

Reply:
done #292